### ISSUE TITLE
Handling HTTP failures gracefully

### ISSUE THREAD

**Issue by schulzrinne** [aws/aws-php-sns-message-validator#11]
https://sns.us-east-1.amazonaws.com/SimpleNotificationService-bb750dd426d95ee9390147a5624348ee.pem): failed to open stream: HTTP request failed! HTTP/1.1 503 Service Unavailable Warning at line 60 in /var/www/html/vendor/aws/aws-php-sns-message-validator/src/MessageValidator.php

There is no way to catch this error.

**Comment by jeremeamia** [comment]
Actually there is, though it's not immediately obvious. But, first, some history...

We added this feature back in version 2 of the SDK. It was a part of the core SDK, and so it made sense to use Guzzle as the HTTP client to fetch the certificates. However, many users complained and said that this was actually the only part of of the SDK they were using, and that they wanted to have it as a separate package. So, in version 3, we separated it into a package with no dependencies on the SDK or Guzzle. That meant using as less robust HTTP client (e.g., `file_get_contents`) to fetch the cert.

So, the error you are getting is being triggered by `file_get_contents()`. All is not lost though. In order to keep this library flexible, we designed the constructor of the `MessageValidator` class to accept a callable to act as the HTTP client. See https://github.com/aws/aws-php-sns-message-validator/blob/master/src/MessageValidator.php#L30-L45. This way you can inject your own function that adds any kind of additional logic you like, including error handling.

In conclusion, you can inject your own error handling logic via a function into the `MessageValidator` to overcome this issue you are observing.

**Comment by schulzrinne** [comment]
Thanks for the ultra-fast response. This is helpful, but I wonder if you have an example of a suitable callable function that works well. Also, any idea why AWS SNS returns 503 messages? Do you know if retrying immediately (within a second or two) is likely to fix the problem?

**Comment by jeskew** [comment]
The callable doesn't need to be anything complicated -- any function that takes a URL as its only argument and returns the response body as a string would work. If you want to use Guzzle 6 so that you can have access to a handler stack, you could write a function that looks a bit like this:

```php
$client = new \GuzzleHttp\Client();
$certClient = function ($url) use ($client) {
    $response = $client->get($url);
    return (string) $response->getBody();
}
```

I'm not sure what caused SNS to return a 503, but I would suspect that transient conditions were to blame and that retrying after a short delay would likely fix the problem.

**Comment by jeskew** [comment]
Closing, as I believe the questions were answered. Please feel free to reopen if you have any questions or concerns.

**Comment by schulzrinne** [comment]
Yes, thanks for your help and suggestions.
